squad-server/plugins/auto-kick-unassigned.js

```diff
@@ -0,0 +1,225 @@
+        default: false
+      }
+    };
+  }
+
+  /**
+   * trackedPlayers[<steam64ID>] = <tracker>
+   *
+   *  <tracker> = {
+   *         player: <playerObj>
+   *       warnings: <int>
+   *      startTime: <Epoch Date>
+   *    warnTimerID: <intervalID>
+   *    kickTimerID: <timeoutID>
+   *  }
+   */
+  constructor(server, options) {
+    super();
+
+    this.server = server;
+    this.options = options;
+
+    this.kickTimeout = options.unassignedTimer * 60 * 1000;
+    this.warningInterval = options.frequencyOfWarnings * 1000;
+    this.gracePeriod = options.roundStartDelay * 60 * 1000;
+
+    this.trackingListUpdateFrequency = 1 * 60 * 1000; // 1min
+    this.cleanUpFrequency = 20 * 60 * 1000; // 20min
+
+    this.betweenRounds = false;
+
+    this.trackedPlayers = {};
+
+    server.on('NEW_GAME', async (info) => {
+      this.betweenRounds = true;
+      this.updateTrackingList();
+      setTimeout(async () => {
+        this.betweenRounds = false;
+      }, this.gracePeriod);
+    });
+
+    server.on('PLAYER_SQUAD_CHANGE', async (player) => {
+      if (player.steamID in this.trackedPlayers && player.squadID !== null) {
+        this.untrackPlayer(player.steamID);
+      }
+    });
+
+    // tracking list update loop
+    setInterval(this.updateTrackingList.bind(this), this.trackingListUpdateFrequency);
+
+    // removes players no longer on the server that may be in trackedPlayers
+    setInterval(() => {
+      for (const steamID of Object.keys(this.trackedPlayers))
+        if (!(steamID in server.players.map((p) => p.steamID))) this.untrackPlayer(steamID);
+    }, this.cleanUpFrequency);
+  }
+
+  runConditions() {
+    // return true; // force run for testing
+    const queueMet =
+      this.options.queueThreshold > 0 &&
+      this.options.queueThreshold < this.server.publicQueue + this.server.reserveQueue;
+    const countMet =
+      this.options.playerCountThreshold > 0 &&
+      this.options.playerCountThreshold < this.server.players.count;
+    const run = !this.betweenRounds && (queueMet || countMet);
+
+    Logger.verbose(
+      'AutoKick',
+      3,
+      `RUN?: ${run} = ${!this.betweenRounds} && (${queueMet} || ${countMet})`
+    );
+    return run;
+  }
+
+  async updateTrackingList(forceUpdate = false) {
+    if (!this.runConditions()) {
+      // clear all tracked players if run conditions are not met.
+      for (const steamID of Object.keys(this.trackedPlayers)) this.untrackPlayer(steamID);
+      return;
+    }
+
+    if (forceUpdate) await this.server.updatePlayerList();
+
+    // loop through players on server and start tracking players not in a squad
+    for (const player of this.server.players) {
+      const isTracked     = player.steamID in this.trackedPlayers;
+      const isUnassigned  = player.squadID === null;
+      const isAdmin       = player.steamID in this.options.adminList.admins;
+      const isWhitelist   = player.steamID in this.options.adminList.whitelist;
```

Comment:
As discussed above, we should manage this info this inside the `SquadServer` class. So maybe add a method `hasAdminPerm(perm, steamID)` to return true/false and then continue in the loop as appropriate.